novawallet/Modules/NetworkNode/Interactor/NetworkNodeBaseInteractor.swift

```diff
@@ -0,0 +1,151 @@
+    
+    init(
+        chainRegistry: any ChainRegistryProtocol,
+        connectionFactory: any ConnectionFactoryProtocol,
+        chainId: ChainModel.Id,
+        repository: AnyDataProviderRepository<ChainModel>,
+        operationQueue: OperationQueue
+    ) {
+        self.chainRegistry = chainRegistry
+        self.connectionFactory = connectionFactory
+        self.chainId = chainId
+        self.repository = repository
+        self.operationQueue = operationQueue
+    }
+    
+    func completeSetup() {
+        subscribeChainChanges()
+    }
+    
+    func setup() {
+        completeSetup()
+    }
+    
+    func connect(
+        to node: ChainNodeModel,
+        chain: ChainModel
+    ) {
+        
+        if let existingNode = findExistingNode(with: node.url, in: chain) {
+            let error = Errors.alreadyExists(nodeName: existingNode.name)
+            basePresenter?.didReceive(error)
+            
+            return
+        }
+        
+        guard wssPredicate.evaluate(with: node.url) else {
+            basePresenter?.didReceive(Errors.wrongFormat)
+            
+            return
+        }
+        
+        currentConnectingNode = node
+        
+        do {
+            currentConnection = try connectionFactory.createConnection(
+                for: node,
+                chain: chain,
+                delegate: self
+            )
+        } catch {
+            basePresenter?.didReceive(Errors.unableToConnect(networkName: chain.name))
+        }
+    }
+    
+    func findExistingNode(
+        with url: String,
+        in chain: ChainModel
+    ) -> ChainNodeModel? {
+        fatalError("Must be overriden by subclass")
+    }
+    
+    func handleConnected() {
+        fatalError("Must be overriden by subclass")
+    }
+}
+
+// MARK: WebSocketEngineDelegate
+
+extension NetworkNodeBaseInteractor: WebSocketEngineDelegate {
+    func webSocketDidSwitchURL(_: AnyObject, newUrl _: URL) {}
+
+    func webSocketDidChangeState(
+        _ connection: AnyObject,
+        from oldState: WebSocketEngine.State,
+        to newState: WebSocketEngine.State
+    ) {
+        guard oldState != newState else { return }
+        
+        DispatchQueue.main.async {
+            guard
+                let chain = self.chainRegistry.getChain(for: self.chainId),
+                let connection = connection as? ChainConnection
+            else {
+                return
+            }
+
+            switch newState {
+            case .notConnected:
+                self.basePresenter?.didReceive(Errors.unableToConnect(networkName: chain.name))
+                self.currentConnection = nil
+            case .waitingReconnection:
+                connection.disconnect(true)
+            case .connected:
+                self.handleConnected()
+            default:
+                break
+            }
+        }
+    }
+}
+
+// MARK: Errors
+
+extension NetworkNodeBaseInteractor {
+    enum Errors: Error {
```

Comment:
So in this case presenter will know about NetworkNodeBaseInteractor.Errors. I would just define NetworkNodeInteractorError. And we later don't need to specify the full type when passing the error to the presenter just, for example, ```basePresenter?.didReceive(.alreadyExists(...))```

Reply:
agree